lib/polymorphic_integer_type/extensions.rb

```diff
@@ -20,37 +20,11 @@ def belongs_to(name, scope = nil, **options)
           _polymorphic_foreign_types << foreign_type
 
           # Required way to dynamically define a class method on the model
-          singleton_class.__send__(:define_method, "#{foreign_type}_mapping") do
+          define_singleton_method("#{foreign_type}_mapping") do
             mapping
           end
 
-          foreign_type_extension = Module.new do
-            define_method foreign_type do
-              t = super()
-              self.class.send("#{foreign_type}_mapping")[t]
-            end
-
-            define_method "#{foreign_type}=" do |klass|
-              mapping = self.class.send("#{foreign_type}_mapping")
-              enum = mapping.key(klass.to_s)
-              if klass.kind_of?(Class) && klass <= ActiveRecord::Base
-                enum ||= mapping.key(klass.polymorphic_name) if klass.respond_to?(:polymorphic_name)
-                enum ||= mapping.key(klass.sti_name)
-                enum ||= mapping.key(klass.base_class.to_s)
-                enum ||= mapping.key(klass.base_class.sti_name)
-              end
-              enum ||= klass if klass != NilClass
-              super(enum)
-            end
-
-            define_method "#{name}=" do |record|
-              super(record)
-              send("#{foreign_type}=", record.class)
-              association(name).loaded!
-            end
-          end
-
-          include(foreign_type_extension)
+          ModuleGenerator.generate_and_include(self, foreign_type, name)
```

Comment:
previously, the define_method blocks that were abstracted would capture the `mapping` variable and could cause really odd behaviour when stubbing it in specs, this ensures that all the stuff it needs it encapsulated in a class and isn't affected by the outside world.